### DISCUSSION TITLE
Typing interaction between `__init_subclass__` and implicit column declaration in 2.0+?

### DISCUSSION THREAD

**Discussion by DanCardin** [sqlalchemy/sqlalchemy#10607] [Usage Questions] [answered]
So the below code snippet is an attempt to allow succinctly implementing behavior that might be common amongst all/many tables.

```python
from datetime import datetime
from sqlalchemy.orm import types, text
from sqlalchemy.orm import DeclarativeBase, Mapped

class Base(DeclarativeBase):
    created_at: Mapped[datetime]

    @classmethod
    def __init_subclass__(cls, *, created_at: bool = False):
        if created_at:
            cls.created_at = mapped_column(
                types.DateTime(timezone=True),
                default=datetime.utcnow,
                server_default=text("CURRENT_TIMESTAMP"),
                nullable=False,
            )

```

Essentially, `class Foo(Base, created_at=True)` should enable this `created_at` column.

Obviously there's mixins.  Or alternatively the `type_annotation_map` feature, might be similarly ergonomic to what i'm attempting to do here. but we had something more akin to the below snippet in 1.4; and i'm attempting to port it to 2.0+.

The **problem** seems to be the interaction between typing, and (i think) the new behavior added into sqlalchemy 2 wherein it infers that you want a column to exist, by annotating the field with `Mapped`.

I'm much less concerned about typing thinking the column exists when it actually doesnt, although if there's no good fix for the below issues, then I suppose I'll have to revert to Mixins

So the class-level annotation is so that `Foo.deleted_at` at query-time is aware of the fact that it should be a column and allows something like `.is_(None)`. Without it, typing complains.

That same class level annotation seems to make sqlalchemy (and thus alembic) think there is an implied column there, and attempts to generate that column whether or not it was enabled for the table in question.

tl;dr, is there a way to indicate to sqlalchemy that the class-level annotation there does not imply that i want a runtime column there, while still making mypy think that there **is** a column there.

**Comment by zzzeek** [comment, marked as answer]
simple answer is put the `Mapped` annotations etc. inside of `TYPE_CHECKING` blocks, Declarative wont see these at runtime.

I guess if you really want to be using `__init_subclass__` to be setting these up and not using a mixin then that's what you'd have to be doing here.

**Reply by DanCardin** [reply]
ahhhhh right, duh!

I dont know that i **want** to be using `__init_subclass__` (although i dont hate the API it yields me), it's more to reduce the diff at the point of upgrade here.
